lighthouse-treemap/app/src/main.js

```diff
@@ -699,61 +705,206 @@ function getStrings(localeMessages) {
+    this._onFileLoad = this._onFileLoad.bind(this);
 
-  injectOptions(options);
+    this._dragAndDrop = new DragAndDrop(this._onFileLoad);
+    this._github = new GithubApi();
 
-  // eslint-disable-next-line no-console
-  console.log('window.__treemapOptions', window.__treemapOptions);
-}
+    document.addEventListener('paste', this._onPaste);
 
-/**
- * @param {string} message
- */
-function showError(message) {
-  document.body.textContent = message;
+    // Hidden file input to trigger manual file selector.
+    const fileInput = TreemapUtil.find('input#hidden-file-input', document);
+    fileInput.addEventListener('change', e => {
+      if (!e.target) {
+        return;
+      }
+
+      const inputTarget = /** @type {HTMLInputElement} */ (e.target);
+      if (inputTarget.files) {
+        this._dragAndDrop.readFile(inputTarget.files[0]).then(str => {
+          this._onFileLoad(str);
+        });
+      }
+      inputTarget.value = '';
+    });
+
+    // A click on the visual placeholder will trigger the hidden file input.
+    const placeholderTarget = TreemapUtil.find('.treemap-placeholder-inner', document);
+    placeholderTarget.addEventListener('click', e => {
+      const target = /** @type {?Element} */ (e.target);
+
+      if (target && target.localName !== 'input' && target.localName !== 'a') {
+        fileInput.click();
+      }
+    });
+  }
+
+  /**
+   * @param {LH.Treemap.Options} options
+   */
+  init(options) {
+    TreemapUtil.find('.treemap-placeholder').classList.add('hidden');
+    TreemapUtil.find('main').classList.remove('hidden');
+
+    const i18n = new I18n(options.lhr.configSettings.locale, {
+      // Set missing renderer strings to default (english) values.
+      ...TreemapUtil.UIStrings,
+      // `strings` is generated in build/build-treemap.js
+      ...getStrings(strings[options.lhr.configSettings.locale]),
+    });
+    TreemapUtil.i18n = i18n;
+
+    // Fill in all i18n data.
+    for (const node of document.querySelectorAll('[data-i18n]')) {
+      // These strings are guaranteed to (at least) have a default English string in TreemapUtil.UIStrings,
+      // so this cannot be undefined as long as `report-ui-features.data-i18n` test passes.
+      const i18nAttr = /** @type {keyof TreemapUtil['UIStrings']} */ (
+        node.getAttribute('data-i18n'));
+      node.textContent = TreemapUtil.i18n.strings[i18nAttr];
+    }
+
+    if (treemapViewer) {
+      TreemapUtil.find('.lh-treemap').innerHTML = '';
+      TreemapUtil.find('.lh-table').innerHTML = '';
+      treemapViewer.abortController.abort();
+    }
+    treemapViewer = new TreemapViewer(options, TreemapUtil.find('div.lh-treemap'));
+
+    injectOptions(options);
+
+    // eslint-disable-next-line no-console
+    console.log('window.__treemapOptions', window.__treemapOptions);
+  }
+
+  /**
+   * @param {unknown} json
+   * @return {LH.Treemap.Options}
+   */
+  convertToOptions(json) {
+    if (json instanceof Object) {
```

Comment:
Interesting I don't think I've really seen `instanceof Object` checks (because cases like `(() => {}) instanceof Object` and typescript gets super weird with `instanceof Object` on unknown), is there a particular situation you like use this over `typeof o === 'object' && o` or just preference?

Reply:
just forgot aboot typeof

Reply:
i don't really care to make this a real type guard function, because it gets complex: https://stackoverflow.com/a/59033939/2788187

so i'm just gonna coerce things here.

Reply:
yeah coercing types is fine, I agree effort:reward isn't there 👍 

This comment was mostly exploratory out of curiosity :)